**Recursively counting words in markdown within a folder**

A question had popped into my mind this week, "How many words have I written in this blog?". As per normal, we'll go into how to use Python to solve this problem. This is an opportunity to get more familiar with `pathlib`, a renowned, enjoyable way to handle pathing within Python programs.

In [1]:
import os
import re
import pathlib
import io
from typing import List
from nbformat import current

A great way to start projects, is to see if it's already been done before. [@gandreadis](https://github.com/gandreadis) over at GitHub had already tackled this problem on the single file basis: [https://github.com/gandreadis/markdown-word-count](https://github.com/gandreadis/markdown-word-count).

This means all we need to do is:
- Recursively loop through a folder and find all the markdown files
- Call [@gandreadis](https://github.com/gandreadis) function to get the count of words for each file
- Sum total words

First of let's bring [@gandreadis](https://github.com/gandreadis) function into our program. It's been modified slightly to take in a file path, read the file contents in a variable, clean up contents and return the length of an array split by spaces.

For example if the sentence (string) is 'How many words have I written in this blog?', when .split is called this returns an array of each word split by a space. Allowing us to calculate how many words, by counting the number of elements in the array.

In [6]:
testSentence = 'How many words have I written in this blog?'
print(testSentence.split())

['How', 'many', 'words', 'have', 'I', 'written', 'in', 'this', 'blog?']


In [5]:
# Source: https://github.com/gandreadis/markdown-word-count
def count_words_in_markdown(filePath: str):
    with open(filePath, 'r', encoding='utf8') as f:
        text = f.read()

    # Comments
    text = re.sub(r'<!--(.*?)-->', '', text, flags=re.MULTILINE)
    # Tabs to spaces
    text = text.replace('\t', '    ')
    # More than 1 space to 4 spaces
    text = re.sub(r'[ ]{2,}', '    ', text)
    # Footnotes
    text = re.sub(r'^\[[^]]*\][^(].*', '', text, flags=re.MULTILINE)
    # Indented blocks of code
    text = re.sub(r'^( {4,}[^-*]).*', '', text, flags=re.MULTILINE)
    # Replace newlines with spaces for uniform handling
    text = text.replace('\n', ' ')
    # Custom header IDs
    text = re.sub(r'{#.*}', '', text)
    # Remove images
    text = re.sub(r'!\[[^\]]*\]\([^)]*\)', '', text)
    # Remove HTML tags
    text = re.sub(r'</?[^>]*>', '', text)
    # Remove special characters
    text = re.sub(r'[#*`~\-–^=<>+|/:]', '', text)
    # Remove footnote references
    text = re.sub(r'\[[0-9]*\]', '', text)
    # Remove enumerations
    text = re.sub(r'[0-9#]*\.', '', text)

    return len(text.split())

We start by defining the 'top' directory to search within. We call this the 'top' folder as we go 'down' inside more folders inside the 'top' folder. Within pathlib, we can access elements within using `glob`. A great resource for testing glob patterns, as they are not very intuitive for me personally is [https://globster.xyz/](https://globster.xyz/). 

The glob pattern elements we will use are:
- `**` - Feature known as globstar. Matches all files and zero or more directories and subdirectories. 
- `*` - Matches any string.
- `?` - Matches single character.
- `.md` - Matches files with a `.md` suffix.

For this particular application, the posts are structured within folders which represent the year they are written so we use the `20??/` glob pattern as to exclude other folders.

In [30]:
# Top directory to search through
topFolder: pathlib.Path = pathlib.Path('C:/Users/jackm/Documents/GitHub/jackmckew.dev/content')

allMarkdown: List  = []

# Iterate through all files using pathlib
for singleFile in topFolder.glob('**/20??/**/*.md'):
    allMarkdown.append(singleFile)

# If you don't want to use glob patterns, can easily use the 'suffix' within the path variable.
# for singleFile in topFolder.glob('**/*'):
#     if singleFile.suffix == '.md':
#         allMarkdown.append(singleFile)

print(len(allMarkdown))

55


We can see that there is 55 markdown files within the folder and all the paths are stored within the list.

All we need to do now is loop over all of them, passing the path into a function defined previously to get the total word count.

In [31]:
totalWordCount: int = 0
for singleFile in allMarkdown:
    totalWordCount += count_words_in_markdown(singleFile)

print(totalWordCount)

31380


There you have it! 31380 words across all the markdown files. In comparison, my engineering thesis for graduating university was 9916 words across 69 pages.

> This currently doesn't include posts that are written within jupyter notebooks, like this one. Later on, we'll go through how to count all the words within markdown cells within jupyter notebooks.